t/api/04-jobs-clone.t

```diff
@@ -0,0 +1,148 @@
+#!/usr/bin/env perl
+# Copyright (C) 2014-2020
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, see <http://www.gnu.org/licenses/>.
+
+use Test::Most;
+
+use FindBin;
+use lib "$FindBin::Bin/lib";
+use OpenQA::Script::CloneJob;
+use OpenQA::Test::Database;
+use Test::Mojo;
+use Test::Warnings ':report_warnings';
+
+OpenQA::Test::Database->new->create();
+
+my $t = Test::Mojo->new('OpenQA::WebAPI');
+# XXX: https://github.com/kraih/mojo/issues/598
+my $app = $t->app;
+$t->ua(
+    OpenQA::Client->new(apikey => 'PERCIVALKEY02', apisecret => 'PERCIVALSECRET02')->ioloop(Mojo::IOLoop->singleton));
+$t->app($app);
+
+my $schema     = $t->app->schema;
+my $products   = $schema->resultset("Products");
+my $testsuites = $schema->resultset("TestSuites");
+my $jobs       = $schema->resultset("Jobs");
+
+my $minimalx = $jobs->find(99926);
+my $clones   = $minimalx->duplicate();
+my $clone    = $jobs->find($clones->{$minimalx->id}->{clone});
+
+isnt($clone->id, $minimalx->id, "is not the same job");
+is($clone->TEST,     "minimalx",  "but is the same test");
+is($clone->priority, 56,          "with the same priority");
+is($minimalx->state, "done",      "original test keeps its state");
+is($clone->state,    "scheduled", "the new job is scheduled");
+
+# Second attempt
+ok($minimalx->can_be_duplicated, "looks cloneable");
+is($minimalx->duplicate, undef, "cannot clone again");
+
+# Reload minimalx from the database
+$minimalx->discard_changes;
+is($minimalx->clone_id,  $clone->id,    "relationship is set");
+is($minimalx->clone->id, $clone->id,    "relationship works");
+is($clone->origin->id,   $minimalx->id, "reverse relationship works");
+
+# After reloading minimalx, it doesn't look cloneable anymore
+ok(!$minimalx->can_be_duplicated, "doesn't look cloneable after reloading");
+is($minimalx->duplicate, undef, "cannot clone after reloading");
+
+# But cloning the clone should be possible after job state change
+$clone->state(OpenQA::Jobs::Constants::CANCELLED);
+$clones = $clone->duplicate({prio => 35});
+my $second = $jobs->find($clones->{$clone->id}->{clone});
+is($second->TEST,     "minimalx", "same test again");
+is($second->priority, 35,         "with adjusted priority");
+
+subtest 'handle settings when posting job' => sub {
```

Comment:
I think this test code is not related with `clone` operation. `Handle settings` happens when creating a job using `job post`. And `clone job` does not handle settings, it just uses the settings (copy from old job) to create a new one. I guess you put this test code here because the next one `do not re-generate settings when cloning job` depends on it.

Reply:
Yeah. They kinda come in tandem.